File: python/databases/db_translator.py
"""Database Translation — dialect-aware DDL emission.

Abstract types render differently per dialect:
  SmallInt → MySQL TINYINT / Postgres SMALLINT / SQLite INTEGER
  Boolean  → MySQL TINYINT(1) / Postgres BOOLEAN / SQLite INTEGER
  Timestamp → MySQL DATETIME / Postgres TIMESTAMP / SQLite TEXT
AUTO_INCREMENT / IDENTITY / AUTOINCREMENT each have their own syntax.
NOW() vs CURRENT_TIMESTAMP for default-timestamp.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from enum import Enum


class Dialect(Enum):
    MYSQL = "mysql"
    POSTGRES = "postgres"
    SQLITE = "sqlite"


class ColumnKind(Enum):
    SMALLINT = "smallint"
    INTEGER = "integer"
    BIGINT = "bigint"
    TEXT = "text"
    VARCHAR = "varchar"
    BOOLEAN = "boolean"
    TIMESTAMP = "timestamp"
    DATE = "date"
    FLOAT = "float"
    DECIMAL = "decimal"


@dataclass
class ColumnType:
    kind: ColumnKind
    length: int = 0             # VARCHAR length
    precision: int = 0          # DECIMAL precision
    scale: int = 0              # DECIMAL scale

    @classmethod
    def small_int(cls) -> ColumnType: return cls(ColumnKind.SMALLINT)
    @classmethod
    def integer(cls) -> ColumnType: return cls(ColumnKind.INTEGER)
    @classmethod
    def big_int(cls) -> ColumnType: return cls(ColumnKind.BIGINT)
    @classmethod
    def text(cls) -> ColumnType: return cls(ColumnKind.TEXT)
    @classmethod
    def varchar(cls, length: int) -> ColumnType: return cls(ColumnKind.VARCHAR, length=length)
    @classmethod
    def boolean(cls) -> ColumnType: return cls(ColumnKind.BOOLEAN)
    @classmethod
    def timestamp(cls) -> ColumnType: return cls(ColumnKind.TIMESTAMP)
    @classmethod
    def date(cls) -> ColumnType: return cls(ColumnKind.DATE)
    @classmethod
    def float_(cls) -> ColumnType: return cls(ColumnKind.FLOAT)
    @classmethod
    def decimal(cls, precision: int, scale: int) -> ColumnType:
        return cls(ColumnKind.DECIMAL, precision=precision, scale=scale)

    def render(self, dialect: Dialect) -> str:
        k = self.kind
        if k == ColumnKind.SMALLINT:
            if dialect == Dialect.MYSQL: return "TINYINT"
            if dialect == Dialect.POSTGRES: return "SMALLINT"
            return "INTEGER"
        if k == ColumnKind.INTEGER:
            return "INTEGER"
        if k == ColumnKind.BIGINT:
            return "INTEGER" if dialect == Dialect.SQLITE else "BIGINT"
        if k == ColumnKind.TEXT:
            return "TEXT"
        if k == ColumnKind.VARCHAR:
            return f"VARCHAR({self.length})"
        if k == ColumnKind.BOOLEAN:
            if dialect == Dialect.MYSQL: return "TINYINT(1)"
            if dialect == Dialect.POSTGRES: return "BOOLEAN"
            return "INTEGER"
        if k == ColumnKind.TIMESTAMP:
            if dialect == Dialect.MYSQL: return "DATETIME"
            if dialect == Dialect.POSTGRES: return "TIMESTAMP"
            return "TEXT"
        if k == ColumnKind.DATE:
            return "TEXT" if dialect == Dialect.SQLITE else "DATE"
        if k == ColumnKind.FLOAT:
            return "REAL"
        if k == ColumnKind.DECIMAL:
            return f"DECIMAL({self.precision},{self.scale})"
        return "UNKNOWN"


class DefaultKind(Enum):
    LITERAL = "literal"
    NOW = "now"


@dataclass
class DefaultValue:
    kind: DefaultKind
    literal: str = ""

    @classmethod
    def of_literal(cls, s: str) -> DefaultValue:
        return cls(DefaultKind.LITERAL, literal=s)

    @classmethod
    def now(cls) -> DefaultValue:
        return cls(DefaultKind.NOW)

    def render(self, dialect: Dialect) -> str:
        if self.kind == DefaultKind.LITERAL:
            return self.literal
        if dialect == Dialect.MYSQL:
            return "NOW()"
        return "CURRENT_TIMESTAMP"


@dataclass
class DbColumn:
    name: str
    ty: ColumnType
    nullable: bool = True
    primary_key: bool = False
    unique: bool = False
    auto_increment: bool = False
    default: DefaultValue | None = None

    def pk(self) -> DbColumn:
        self.primary_key = True
        self.nullable = False
        return self

    def not_null(self) -> DbColumn:
        self.nullable = False
        return self

    def make_unique(self) -> DbColumn:
        self.unique = True
        return self

    def auto_inc(self) -> DbColumn:
        self.auto_increment = True
        return self

    def with_default(self, d: DefaultValue) -> DbColumn:
        self.default = d
        return self


@dataclass
class Table:
    name: str
    columns: list[DbColumn] = field(default_factory=list)


def _render_auto_increment(dialect: Dialect) -> str:
    return {
        Dialect.MYSQL: "AUTO_INCREMENT",
        Dialect.POSTGRES: "GENERATED ALWAYS AS IDENTITY",
        Dialect.SQLITE: "AUTOINCREMENT",
    }[dialect]


def _render_column(col: DbColumn, dialect: Dialect) -> str:
    parts = [col.name, col.ty.render(dialect)]
    if col.primary_key:
        parts.append("PRIMARY KEY")
    if not col.nullable and not col.primary_key:
        parts.append("NOT NULL")
    if col.unique and not col.primary_key:
        parts.append("UNIQUE")
    if col.auto_increment:
        if dialect == Dialect.SQLITE and col.primary_key:
            parts.append("AUTOINCREMENT")
        elif dialect != Dialect.SQLITE:
            parts.append(_render_auto_increment(dialect))
    if col.default is not None:
        parts.append(f"DEFAULT {col.default.render(dialect)}")
    return " ".join(parts)


def emit_create_table(table: Table, dialect: Dialect) -> str:
    lines = [f"CREATE TABLE {table.name} ("]
    cols = [f"  {_render_column(c, dialect)}" for c in table.columns]
    lines.append(",\n".join(cols))
    lines.append(");")
    return "\n".join(lines) + "\n"


def emit_schema(tables: list[Table], dialect: Dialect) -> str:
    return "".join(emit_create_table(t, dialect) for t in tables)


def demo() -> None:
    users = Table(
        name="users",
        columns=[
            DbColumn("id", ColumnType.integer()).pk().auto_inc(),
            DbColumn("email", ColumnType.varchar(255)).not_null().make_unique(),
            DbColumn("name", ColumnType.text()).not_null(),
            DbColumn("active", ColumnType.boolean()).not_null()
                    .with_default(DefaultValue.of_literal("1")),
            DbColumn("created_at", ColumnType.timestamp()).not_null()
                    .with_default(DefaultValue.now()),
        ],
    )
    for dialect in [Dialect.MYSQL, Dialect.POSTGRES, Dialect.SQLITE]:
        print(f"=== {dialect.value} ===")
        print(emit_create_table(users, dialect))


if __name__ == "__main__":
    demo()

File: rust/databases/db_translator.rs
//! Database Translation — dialect-aware DDL emission.
//!
//! Twelfth Databases project. Every team that's migrated between
//! MySQL and Postgres (or SQLite, or SQL Server) has hit this: the
//! *logical* schema is the same, but the *syntax* differs per dialect.
//! TINYINT becomes SMALLINT. AUTO_INCREMENT becomes SERIAL. NOW()
//! becomes CURRENT_TIMESTAMP. G112 captures the translation rules.
//!
//! The distinctive moves:
//!   * **Abstract type** — `ColumnType` (Integer, BigInt, Text,
//!     Varchar(n), Boolean, Timestamp, etc.) that every dialect
//!     renders differently.
//!   * **Dialect-specific rendering functions** for types, auto-
//!     increment, default-now.
//!   * **DDL emission** — `CREATE TABLE` statement per table, with
//!     primary keys, not-nulls, uniques, defaults, and auto-increment.
//!   * **Cross-dialect portability** — the same logical schema emits
//!     valid DDL for any supported dialect.

use std::fmt::Write;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Dialect { MySql, Postgres, Sqlite }

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ColumnType {
    SmallInt,
    Integer,
    BigInt,
    Text,
    Varchar(u32),
    Boolean,
    Timestamp,
    Date,
    Float,
    Decimal(u32, u32),  // precision, scale
}

impl ColumnType {
    pub fn render(&self, dialect: Dialect) -> String {
        match (self, dialect) {
            (ColumnType::SmallInt, Dialect::MySql) => "TINYINT".into(),
            (ColumnType::SmallInt, Dialect::Postgres) => "SMALLINT".into(),
            (ColumnType::SmallInt, Dialect::Sqlite) => "INTEGER".into(),
            (ColumnType::Integer, _) => "INTEGER".into(),
            (ColumnType::BigInt, Dialect::Sqlite) => "INTEGER".into(),
            (ColumnType::BigInt, _) => "BIGINT".into(),
            (ColumnType::Text, _) => "TEXT".into(),
            (ColumnType::Varchar(n), _) => format!("VARCHAR({n})"),
            (ColumnType::Boolean, Dialect::MySql) => "TINYINT(1)".into(),
            (ColumnType::Boolean, Dialect::Postgres) => "BOOLEAN".into(),
            (ColumnType::Boolean, Dialect::Sqlite) => "INTEGER".into(),
            (ColumnType::Timestamp, Dialect::MySql) => "DATETIME".into(),
            (ColumnType::Timestamp, Dialect::Postgres) => "TIMESTAMP".into(),
            (ColumnType::Timestamp, Dialect::Sqlite) => "TEXT".into(),
            (ColumnType::Date, Dialect::Sqlite) => "TEXT".into(),
            (ColumnType::Date, _) => "DATE".into(),
            (ColumnType::Float, _) => "REAL".into(),
            (ColumnType::Decimal(p, s), _) => format!("DECIMAL({p},{s})"),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum DefaultValue {
    /// SQL literal, e.g. "'active'" (caller provides quoting).
    Literal(String),
    /// The current timestamp; dialect-specific.
    Now,
}

impl DefaultValue {
    pub fn render(&self, dialect: Dialect) -> String {
        match self {
            DefaultValue::Literal(s) => s.clone(),
            DefaultValue::Now => match dialect {
                Dialect::MySql => "NOW()".into(),
                Dialect::Postgres | Dialect::Sqlite => "CURRENT_TIMESTAMP".into(),
            },
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DbColumn {
    pub name: String,
    pub ty: ColumnType,
    pub nullable: bool,
    pub primary_key: bool,
    pub unique: bool,
    pub auto_increment: bool,
    pub default: Option<DefaultValue>,
}

impl DbColumn {
    pub fn new(name: &str, ty: ColumnType) -> Self {
        Self {
            name: name.into(), ty,
            nullable: true, primary_key: false, unique: false,
            auto_increment: false, default: None,
        }
    }
    pub fn pk(mut self) -> Self { self.primary_key = true; self.nullable = false; self }
    pub fn not_null(mut self) -> Self { self.nullable = false; self }
    pub fn unique(mut self) -> Self { self.unique = true; self }
    pub fn auto_increment(mut self) -> Self { self.auto_increment = true; self }
    pub fn default(mut self, d: DefaultValue) -> Self { self.default = Some(d); self }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DbTranslatorTable {
    pub name: String,
    pub columns: Vec<DbColumn>,
}

fn render_auto_increment(dialect: Dialect) -> &'static str {
    match dialect {
        Dialect::MySql => "AUTO_INCREMENT",
        Dialect::Postgres => "GENERATED ALWAYS AS IDENTITY",
        Dialect::Sqlite => "AUTOINCREMENT",
    }
}

fn render_column(col: &DbColumn, dialect: Dialect) -> String {
    let mut out = String::new();
    // For Postgres with auto_increment on integer types, use SERIAL-family.
    // Actually the modern Postgres approach is IDENTITY; emit that.
    write!(&mut out, "{} {}", col.name, col.ty.render(dialect)).unwrap();
    if col.primary_key {
        out.push_str(" PRIMARY KEY");
    }
    if !col.nullable && !col.primary_key {
        out.push_str(" NOT NULL");
    }
    if col.unique && !col.primary_key {
        out.push_str(" UNIQUE");
    }
    if col.auto_increment {
        // Sqlite requires "INTEGER PRIMARY KEY AUTOINCREMENT".
        if dialect == Dialect::Sqlite && col.primary_key {
            out.push_str(" AUTOINCREMENT");
        } else if dialect != Dialect::Sqlite {
            out.push(' ');
            out.push_str(render_auto_increment(dialect));
        }
    }
    if let Some(d) = &col.default {
        out.push_str(&format!(" DEFAULT {}", d.render(dialect)));
    }
    out
}

/// Emit a `CREATE TABLE` statement for the given dialect.
pub fn emit_create_table(table: &DbTranslatorTable, dialect: Dialect) -> String {
    let mut out = String::new();
    writeln!(&mut out, "CREATE TABLE {} (", table.name).unwrap();
    let cols: Vec<String> = table.columns.iter()
        .map(|c| format!("  {}", render_column(c, dialect)))
        .collect();
    out.push_str(&cols.join(",\n"));
    out.push('\n');
    out.push_str(");\n");
    out
}

/// Emit CREATE TABLE for all tables in a given dialect, concatenated.
pub fn emit_schema(tables: &[DbTranslatorTable], dialect: Dialect) -> String {
    let mut out = String::new();
    for table in tables {
        out.push_str(&emit_create_table(table, dialect));
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    fn users_table() -> DbTranslatorTable {
        DbTranslatorTable {
            name: "users".into(),
            columns: vec![
                DbColumn::new("id", ColumnType::Integer).pk().auto_increment(),
                DbColumn::new("email", ColumnType::Varchar(255)).not_null().unique(),
                DbColumn::new("name", ColumnType::Text).not_null(),
                DbColumn::new("active", ColumnType::Boolean).not_null()
                    .default(DefaultValue::Literal("1".into())),
                DbColumn::new("created_at", ColumnType::Timestamp).not_null()
                    .default(DefaultValue::Now),
            ],
        }
    }

    #[test]
    fn tinyint_is_mysql_specific() {
        assert_eq!(ColumnType::SmallInt.render(Dialect::MySql), "TINYINT");
        assert_eq!(ColumnType::SmallInt.render(Dialect::Postgres), "SMALLINT");
    }

    #[test]
    fn boolean_differs_per_dialect() {
        assert_eq!(ColumnType::Boolean.render(Dialect::MySql), "TINYINT(1)");
        assert_eq!(ColumnType::Boolean.render(Dialect::Postgres), "BOOLEAN");
        assert_eq!(ColumnType::Boolean.render(Dialect::Sqlite), "INTEGER");
    }

    #[test]
    fn timestamp_differs_per_dialect() {
        assert_eq!(ColumnType::Timestamp.render(Dialect::MySql), "DATETIME");
        assert_eq!(ColumnType::Timestamp.render(Dialect::Postgres), "TIMESTAMP");
        assert_eq!(ColumnType::Timestamp.render(Dialect::Sqlite), "TEXT");
    }

    #[test]
    fn varchar_preserves_length() {
        assert_eq!(ColumnType::Varchar(255).render(Dialect::MySql), "VARCHAR(255)");
    }

    #[test]
    fn decimal_preserves_precision_scale() {
        assert_eq!(ColumnType::Decimal(10, 2).render(Dialect::Postgres), "DECIMAL(10,2)");
    }

    #[test]
    fn now_default_differs_per_dialect() {
        assert_eq!(DefaultValue::Now.render(Dialect::MySql), "NOW()");
        assert_eq!(DefaultValue::Now.render(Dialect::Postgres), "CURRENT_TIMESTAMP");
    }

    #[test]
    fn mysql_create_table_uses_auto_increment() {
        let ddl = emit_create_table(&users_table(), Dialect::MySql);
        assert!(ddl.contains("AUTO_INCREMENT"));
        assert!(ddl.contains("NOW()"));
        assert!(ddl.contains("DATETIME"));
        assert!(ddl.contains("TINYINT(1)"));
    }

    #[test]
    fn postgres_create_table_uses_identity() {
        let ddl = emit_create_table(&users_table(), Dialect::Postgres);
        assert!(ddl.contains("GENERATED ALWAYS AS IDENTITY"));
        assert!(ddl.contains("CURRENT_TIMESTAMP"));
        assert!(ddl.contains("TIMESTAMP"));
        assert!(ddl.contains("BOOLEAN"));
    }

    #[test]
    fn sqlite_create_table_uses_autoincrement() {
        let ddl = emit_create_table(&users_table(), Dialect::Sqlite);
        assert!(ddl.contains("AUTOINCREMENT"));
        // Sqlite flattens booleans and timestamps to its native types.
        assert!(ddl.contains("INTEGER"));
    }

    #[test]
    fn primary_key_implies_not_null() {
        let table = DbTranslatorTable {
            name: "t".into(),
            columns: vec![
                DbColumn::new("id", ColumnType::Integer).pk(),
                DbColumn::new("name", ColumnType::Text).not_null(),
            ],
        };
        let ddl = emit_create_table(&table, Dialect::Postgres);
        // PK line has "PRIMARY KEY" but not separate "NOT NULL".
        let pk_line: &str = ddl.lines().find(|l| l.contains("id ")).unwrap();
        assert!(pk_line.contains("PRIMARY KEY"));
        assert!(!pk_line.contains("NOT NULL"));
    }

    #[test]
    fn unique_constraint_emitted() {
        let ddl = emit_create_table(&users_table(), Dialect::Postgres);
        assert!(ddl.contains("email VARCHAR(255) NOT NULL UNIQUE"));
    }

    #[test]
    fn default_literal_passes_through() {
        let ddl = emit_create_table(&users_table(), Dialect::Postgres);
        assert!(ddl.contains("DEFAULT 1"));
    }

    #[test]
    fn emit_schema_concatenates_tables() {
        let tables = vec![
            users_table(),
            DbTranslatorTable {
                name: "posts".into(),
                columns: vec![
                    DbColumn::new("id", ColumnType::Integer).pk().auto_increment(),
                    DbColumn::new("title", ColumnType::Text).not_null(),
                ],
            },
        ];
        let ddl = emit_schema(&tables, Dialect::Postgres);
        assert!(ddl.contains("CREATE TABLE users"));
        assert!(ddl.contains("CREATE TABLE posts"));
    }

    #[test]
    fn same_schema_emits_different_ddl_per_dialect() {
        let mysql = emit_create_table(&users_table(), Dialect::MySql);
        let pg = emit_create_table(&users_table(), Dialect::Postgres);
        assert_ne!(mysql, pg);
    }

    #[test]
    fn bigint_is_integer_in_sqlite() {
        assert_eq!(ColumnType::BigInt.render(Dialect::Sqlite), "INTEGER");
        assert_eq!(ColumnType::BigInt.render(Dialect::Postgres), "BIGINT");
    }

    #[test]
    fn date_is_text_in_sqlite() {
        assert_eq!(ColumnType::Date.render(Dialect::Sqlite), "TEXT");
        assert_eq!(ColumnType::Date.render(Dialect::Postgres), "DATE");
    }
}
